### PR TITLE
feat(inapp): interaction events APIs

### PR DESCRIPTION
**Pull request by Samaritan1011001**
Successor to https://github.com/aws-amplify/amplify-js/pull/12231

#### Description of changes

This PR adds the following APIs along with their unit tests, 
- onMessageReceived
- onMessageDisplayed
- onMessageDismissed
- onMessageActionTaken
- notifyInteractionEvents

#### Description of how you validated changes
Added callback handlers using the APIs mentioned above and calling dispatchEvent triggers them.

#### Checklist

- [x] PR description included
- [x] `yarn test` passes
- [x] Tests are [changed or added](https://github.com/aws-amplify/amplify-js/blob/main/CONTRIBUTING.md#steps-towards-contributions)
- [ ] Relevant documentation is changed or added (and PR referenced)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/notifications/src/inAppMessaging/providers/pinpoint/types/types.ts [hunk 1]
```diff
@@ -30,3 +30,5 @@ export enum PinpointMessageEvent {
 export type InAppMessageConflictHandler = (
 	messages: InAppMessage[]
 ) => InAppMessage;
+
+export type OnMessageInteractionEventHandler = (message: InAppMessage) => any;
```
Comment (HuiSF) [inline packages/notifications/src/inAppMessaging/providers/pinpoint/types/types.ts:34]: nit: event handler functions should return `void`

File: packages/notifications/src/inAppMessaging/providers/pinpoint/apis/notifyMessageInteraction.ts [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+// SPDX-License-Identifier: Apache-2.0
+
+import { notifyEventListeners } from '../../../../common';
+import { NotifyMessageInteractionInput } from '../types/inputs';
+
+/**
+ * Notifies the respective listener of the specified type with the message given.
+ *
+ * @param {NotifyMessageInteractionInput} input - The input object that holds the type and message.
+ *
```
Comment (HuiSF) [inline packages/notifications/src/inAppMessaging/providers/pinpoint/apis/notifyMessageInteraction.ts:11]: nit: extra line

File: packages/notifications/src/common/eventListeners/index.ts [hunk 3]
```diff
@@ -40,5 +40,7 @@ export const addEventListener = <EventHandler extends Function>(
 		},
 	};
 	eventListeners[type].add(listener);
-	return listener;
+	return {
+		remove: listener.remove,
```
Comment (HuiSF) [inline packages/notifications/src/common/eventListeners/index.ts:44]: nit: I'd do `remove: () => listener.remove()` here not worrying about object binding

File: packages/notifications/src/common/eventListeners/index.ts [hunk 4]
```diff
@@ -40,5 +40,7 @@ export const addEventListener = <EventHandler extends Function>(
 		},
 	};
 	eventListeners[type].add(listener);
-	return listener;
```
Comment (jimblanc) [inline packages/notifications/src/common/eventListeners/index.ts:43]: Stupid question, but why don't we need `handleEvent` anymore in the response?

File: packages/notifications/src/inAppMessaging/providers/pinpoint/apis/notifyMessageInteraction.ts [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+ * @returns {void}
+ *
+ * @example
+ * ```ts
+ * onMessageRecieved((message) => {
+ *   // show end users the inApp message and then notify the event
```
Comment (jimblanc) [inline packages/notifications/src/inAppMessaging/providers/pinpoint/apis/notifyMessageInteraction.ts:17]:
```suggestion
 *   // Show end users the In-App message and notify event listeners
```

File: packages/notifications/src/inAppMessaging/providers/pinpoint/apis/onMessageActionTaken.ts [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+
+import { addEventListener } from '../../../../common';
+import { OnMessageActionTakenInput } from '../types/inputs';
+import { OnMessageActionTakenOutput } from '../types/outputs';
+
+/**
+ * Registers a callback handler that is fired upon a message action taken event
+ *
+ * @param {OnMessageActionTakenInput} input - The input object that holds the callback handler.
+ *
```
Comment (jimblanc) [inline packages/notifications/src/inAppMessaging/providers/pinpoint/apis/onMessageActionTaken.ts:12]:
```suggestion
```

File: packages/notifications/src/inAppMessaging/providers/pinpoint/apis/onMessageActionTaken.ts [hunk 7]
```diff
@@ -0,0 +1,27 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+import { addEventListener } from '../../../../common';
+import { OnMessageActionTakenInput } from '../types/inputs';
+import { OnMessageActionTakenOutput } from '../types/outputs';
+
+/**
+ * Registers a callback handler that is fired upon a message action taken event
```
Comment (jimblanc) [inline packages/notifications/src/inAppMessaging/providers/pinpoint/apis/onMessageActionTaken.ts:9]:
```suggestion
 * Registers a callback that will be invoked on `messageActionTaken` events.
```

File: packages/notifications/src/inAppMessaging/providers/pinpoint/apis/onMessageActionTaken.ts [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+ *
+ * @param {OnMessageActionTakenInput} input - The input object that holds the callback handler.
+ *
+ *
+ * @returns {OnMessageActionTakenOutput} - An object that holds a remove method to stop listening to events.
+ *
+ * @example
+ * ```ts
+ * onMessageActionTaken((message) => {
+ *   // use the message
```
Comment (jimblanc) [inline packages/notifications/src/inAppMessaging/providers/pinpoint/apis/onMessageActionTaken.ts:19]: Nit: Could we add a more representative example, e.g. accessing a field in the message

File: packages/notifications/src/inAppMessaging/providers/pinpoint/apis/onMessageDismissed.ts [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+
+import { addEventListener } from '../../../../common';
+import { OnMessageDismissedOutput } from '../types/outputs';
+import { OnMessageDismissedInput } from '../types/inputs';
+
+/**
+ * Registers a callback handler that is fired upon a message dismissed event
+ *
+ * @param {OnMessageDismissedInput} input - The input object that holds the callback handler.
+ *
```
Comment (jimblanc) [inline packages/notifications/src/inAppMessaging/providers/pinpoint/apis/onMessageDismissed.ts:12]:
```suggestion
```

File: packages/notifications/src/inAppMessaging/providers/pinpoint/apis/onMessageDismissed.ts [hunk 10]
```diff
@@ -0,0 +1,27 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+import { addEventListener } from '../../../../common';
+import { OnMessageDismissedOutput } from '../types/outputs';
+import { OnMessageDismissedInput } from '../types/inputs';
+
+/**
+ * Registers a callback handler that is fired upon a message dismissed event
```
Comment (jimblanc) [inline packages/notifications/src/inAppMessaging/providers/pinpoint/apis/onMessageDismissed.ts:9]:
```suggestion
 * Registers a callback that will be invoked on `messageDismissed` events.
```

File: packages/notifications/src/inAppMessaging/providers/pinpoint/apis/onMessageDisplayed.ts [hunk 11]
```diff
@@ -0,0 +1,27 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+import { addEventListener } from '../../../../common';
+import { OnMessageDisplayedOutput } from '../types/outputs';
+import { OnMessageDisplayedInput } from '../types/inputs';
+
+/**
+ * Registers a callback handler that is fired upon a message displayed event
```
Comment (jimblanc) [inline packages/notifications/src/inAppMessaging/providers/pinpoint/apis/onMessageDisplayed.ts:9]:
```suggestion
 * Registers a callback that will be invoked on `messageDisplayed` events.
```

File: packages/notifications/src/inAppMessaging/providers/pinpoint/apis/onMessageDisplayed.ts [hunk 12]
```diff
@@ -0,0 +1,27 @@
...
+ *
+ * @param {OnMessageDisplayedInput} input - The input object that holds the callback handler.
+ *
```
Comment (jimblanc) [inline packages/notifications/src/inAppMessaging/providers/pinpoint/apis/onMessageDisplayed.ts:12]:
```suggestion
```

File: packages/notifications/src/inAppMessaging/providers/pinpoint/apis/onMessageReceived.ts [hunk 13]
```diff
@@ -0,0 +1,27 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+import { addEventListener } from '../../../../common';
+import { OnMessageReceivedInput } from '../types/inputs';
+import { OnMessageReceivedOutput } from '../types/outputs';
+
+/**
+ * Registers a callback handler that is fired upon a message received event
```
Comment (jimblanc) [inline packages/notifications/src/inAppMessaging/providers/pinpoint/apis/onMessageReceived.ts:9]:
```suggestion
 * Registers a callback that will be invoked on `messageReceived` events.
```

File: packages/notifications/src/inAppMessaging/providers/pinpoint/apis/onMessageReceived.ts [hunk 14]
```diff
@@ -0,0 +1,27 @@
...
+ *
+ * @param {OnMessageReceivedInput} input - The input object that holds the callback handler.
+ *
```
Comment (jimblanc) [inline packages/notifications/src/inAppMessaging/providers/pinpoint/apis/onMessageReceived.ts:12]:
```suggestion
```

File: packages/notifications/src/common/eventListeners/types.ts [hunk 15]
```diff
@@ -10,3 +10,7 @@ export interface EventListener<EventHandler extends Function> {
 }
 
 export type EventType = InAppMessageInteractionEvent | PushNotificationEvent;
+
+export type EventListenerRemover = {
```
Comment (jimblanc) [inline packages/notifications/src/common/eventListeners/types.ts:14]: Do we still need the `EventListener` type above?

File: packages/notifications/src/inAppMessaging/index.ts [hunk 16]
```diff
@@ -6,4 +6,9 @@ export {
 	syncMessages,
 	dispatchEvent,
 	setConflictHandler,
+	onMessageReceived,
+	onMessageDisplayed,
+	onMessageDismissed,
+	onMessageActionTaken,
+	notifyMessageInteraction,
 } from './providers/pinpoint';
```
Comment (jimblanc) [inline packages/notifications/src/inAppMessaging/index.ts:14]: Could you please add an export test to `exports.test.ts` to track the export surface?

**Review by jimblanc — COMMENTED** [review]
Looks pretty good, some nits & questions

File: packages/notifications/src/inAppMessaging/index.ts (hunk 16 above)
Comment (jimblanc) [reply packages/notifications/src/inAppMessaging/index.ts:14]: Actually looks like it already exists & just needs to be updated

File: packages/notifications/src/common/eventListeners/index.ts (hunk 4 above)
Comment (cshfang) [reply packages/notifications/src/common/eventListeners/index.ts:43]: I think we probably never needed to expose it as an API on the listener itself as long as the module held a reference to the callback

File: packages/notifications/src/inAppMessaging/providers/pinpoint/apis/setConflictHandler.ts [hunk 17]
```diff
@@ -11,8 +11,7 @@ export let conflictHandler: InAppMessageConflictHandler =
  * Set a conflict handler that will be used to resolve conflicts that may emerge
  * when matching events with synced messages.
  * @remark
- * The conflict handler is not persisted between sessions
- * and needs to be called before dispatching an event to have any effect.
+ * The conflict handler is not persisted across app restarts and needs to be called before dispatching an event.
```
Comment (cshfang) [inline packages/notifications/src/inAppMessaging/providers/pinpoint/apis/setConflictHandler.ts:14]:
```suggestion
 * The conflict handler is not persisted across app restarts and so must be set again before dispatching an event for any custom handling to take effect.
```

File: packages/notifications/src/common/eventListeners/index.ts (hunk 4 above)
Comment (Samaritan1011001) [reply packages/notifications/src/common/eventListeners/index.ts:43]: Yeah, it's only needed internally. It's the same callback handler that is given to the API, doesn't make sense to return it in the response.

File: packages/notifications/src/inAppMessaging/providers/pinpoint/apis/onMessageActionTaken.ts (hunk 8 above)
Comment (Samaritan1011001) [reply packages/notifications/src/inAppMessaging/providers/pinpoint/apis/onMessageActionTaken.ts:19]: Could only think of a console.log, let me know if have anything particular in mind

File: packages/notifications/src/common/eventListeners/types.ts (hunk 15 above)
Comment (Samaritan1011001) [reply packages/notifications/src/common/eventListeners/types.ts:14]: Yeah, still internally used to pair the handler and remove methods. This is only the return type

File: packages/notifications/src/inAppMessaging/providers/pinpoint/apis/notifyMessageInteraction.ts [hunk 18]
```diff
@@ -0,0 +1,27 @@
...
+ *   notifyMessageInteraction({ type: "messageDisplayed", message });
```
[the remark this answers is not in the document]
Comment (Samaritan1011001) [reply packages/notifications/src/inAppMessaging/providers/pinpoint/apis/notifyMessageInteraction.ts:18]: Hadn't observed, don't have a preference. Can change it to single quotes

**Review by jimblanc — APPROVED** [review]

**Review by cshfang — APPROVED** [review]

**Pull request merged**